src/utils/graph/layout.js

```diff
@@ -0,0 +1,319 @@
+      // Find the minimal target position that separates both nodes
+      const sourceDelta = co.edgeA.sourceNode.x - co.edgeB.sourceNode.x;
+      const targetDelta = co.edgeA.targetNode.x - co.edgeB.targetNode.x;
+      return sourceDelta + targetDelta < 0 ? -co.basisX : co.basisX;
+    },
+    strength: co => 1 / co.basisX,
+    weightA: () => 0.5,
+    weightB: () => 0.5
+  };
+
+  // For every pair of edges
+  for (let i = 0; i < edges.length; i += 1) {
+    const edgeA = edges[i];
+
+    for (let j = i + 1; j < edges.length; j += 1) {
+      const edgeB = edges[j];
+
+      // Add crossing constraint between edge source nodes, where different
+      if (edgeA.source !== edgeB.source) {
+        crossingConstraints.push({
+          ...crossingConstraint,
+          a: edgeA.sourceNode,
+          b: edgeB.sourceNode,
+          edgeA,
+          edgeB
+        });
+      }
+
+      // Add crossing constraint between edge target nodes, where different
+      if (edgeA.target !== edgeB.target) {
+        crossingConstraints.push({
+          ...crossingConstraint,
+          a: edgeA.targetNode,
+          b: edgeB.targetNode,
+          edgeA,
+          edgeB
+        });
+      }
+    }
+  }
+
+  // Constraints in X to minimise edge length thereby prioritising straight parallel edges in Y
+  for (const edge of edges) {
+    const parallelConstraint = {
+      a: edge.sourceNode,
+      b: edge.targetNode,
+      key: 'x',
+      operator: equalTo,
+      target: () => 0,
+      // Lower degree nodes can be moved more freely than higher
+      strength: co =>
+        1 / Math.max(1, 0.5 * (co.a.targets.length + co.b.sources.length)),
+      weightA: () => 0.5,
+      weightB: () => 0.5
+    };
+
+    parallelConstraints.push(parallelConstraint);
+
+    // Collect edges connected to single-degree nodes at either end
+    if (
+      edge.sourceNode.targets.length === 1 ||
+      edge.targetNode.sources.length === 1
+    ) {
+      parallelSingleConstraints.push(parallelConstraint);
+    }
+
+    // Collect edges connected to single-degree at both ends
+    if (
+      edge.sourceNode.targets.length === 1 &&
+      edge.targetNode.sources.length === 1
+    ) {
+      parallelDoubleConstraints.push(parallelConstraint);
+    }
+  }
+
+  // Solving loop for constraints in X
+  for (let i = 0; i < 20; i += 1) {
+    // Minimise crossing
+    solve(crossingConstraints, 1);
+
+    // Minimise edge length
+    solve(parallelConstraints, 1);
+
+    // Minimise edge length specifically for low-degree edges more strongly
+    solve(parallelSingleConstraints, 10);
+    solve(parallelDoubleConstraints, 10);
+
+    // Clear separation constraints from previous iteration
+    separationConstraints.length = 0;
+
+    // For each row
+    for (let l = 0; l < rows.length; l += 1) {
+      const rowNodes = rows[l];
+
+      // Sort rows in order of X position. Break ties with ids for stability
+      rowNodes.sort((a, b) => compare(a.x, b.x, a.id, b.id));
```

Comment:
Necessary as it stands. But I think this part should be complexity `<= O(sort(nodes))`, since `flat(rows) === nodes`, so not relatively that bad.

I've not looked at optimising anywhere much yet as overall as performance seems sufficient for an initial flagged release?

When I get on to this I'm sure there will be some simple improvements to find first, then after that the bigger wins will probably take some special techniques.